x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/Case.java

```diff
@@ -0,0 +1,159 @@
+                return new TypeResolution(format(null, "result of [{}] must be [{}], found value [{}] type [{}]",
+                    conditional.sourceText(),
+                    expectedResultDataType.typeName,
+                    Expressions.name(conditional.result()),
+                    conditional.dataType().typeName));
+            }
+        }
+
+        if (DataTypes.areTypesCompatible(dataType, defaultElse.dataType()) == false) {
+            return new TypeResolution(format(null, "ELSE clause of [{}] must be [{}], found value [{}] type [{}]",
+                defaultElse.sourceText(),
+                expectedResultDataType.typeName,
+                Expressions.name(defaultElse),
+                defaultElse.dataType().typeName));
+        }
+
+        return TypeResolution.TYPE_RESOLVED;
+    }
+
+    @Override
+    public boolean foldable() {
+        return defaultElse.foldable() && (conditions.isEmpty() ||
+            (conditions.size() == 1 && conditions.get(0).condition().foldable() && conditions.get(0).result().foldable()));
+    }
+
+    @Override
+    public Object fold() {
+        if (conditions.isEmpty()) {
+            return defaultElse.fold();
+        }
+        if (conditions.get(0).condition().fold() == Boolean.TRUE) {
+            return conditions.get(0).result().fold();
+        }
+        return defaultElse.fold();
+    }
+
+    @Override
+    protected Pipe makePipe() {
+        List<Pipe> pipes = new ArrayList<>(conditions.size() + 1);
+        for (IfConditional ifConditional : conditions) {
+            pipes.add(Expressions.pipe(ifConditional.condition()));
+            pipes.add(Expressions.pipe(ifConditional.result()));
+        }
+        pipes.add(Expressions.pipe(defaultElse));
+        return new CasePipe(source(), this, pipes);
+    }
+
+    @Override
+    public ScriptTemplate asScript() {
+        List<ScriptTemplate> templates = new ArrayList<>();
+        for (IfConditional ifConditional : conditions) {
+            templates.add(asScript(ifConditional.condition()));
+            templates.add(asScript(ifConditional.result()));
+        }
+        templates.add(asScript(defaultElse));
+
+        StringJoiner template = new StringJoiner(",", "{sql}.caseFunction([", "])");
+        ParamsBuilder params = paramsBuilder();
+
+        for (ScriptTemplate scriptTemplate : templates) {
+            template.add(scriptTemplate.template());
+            params.script(scriptTemplate.params());
+        }
+
+        return new ScriptTemplate(formatTemplate(template.toString()), params.build(), dataType());
+    }
+
+    private void setDataType() {
+        if (conditions.isEmpty()) {
+            dataType = defaultElse().dataType();
+        } else {
+            dataType = conditions.get(0).dataType();
+        }
+
+        for (IfConditional conditional : conditions) {
+            dataType = DataTypeConversion.commonType(dataType, conditional.dataType());
```

Comment:
In case `conditions.isEmpty() == false` the first `commonType` check is with itself, isn't it? Maybe worth doing a `for` with an index that starts either at `0` or `1` depending on the previous `isEmpty()` check?

Reply:
Indeed, I would just start with the defaultElse() type and then iterate over the conditions - this avoid repetition.